ecrecover/src/lib.rs

```diff
@@ -0,0 +1,202 @@
+        // Same thing as the geth test, with an invalid '-1' recovery id
+        let input = FromHex::from_hex(&"38d18acb67d25c8bb9942764b62f18e17054f66a817bd4295423adf9ed98873e00000000000000000000000000000000000000000000000000000000000000ff38d18acb67d25c8bb9942764b62f18e17054f66a817bd4295423adf9ed98873e789d1dd423d25f0772d2748d60f7e4b81bb14d086eba8e8e8efb6dcff8a4ae02"[..])
+            .unwrap();
+        let expected = vec![0u8; 0];
+
+        let output = ecrecover(&input).unwrap();
+
+        assert_eq!(expected, output);
+
+        // Same thing as the geth test, with an invalid 'infinite' recovery id
+        let input = FromHex::from_hex(&"38d18acb67d25c8bb9942764b62f18e17054f66a817bd4295423adf9ed98873e000000000000000000000000000000000000000000000000000000000000001e38d18acb67d25c8bb9942764b62f18e17054f66a817bd4295423adf9ed98873e789d1dd423d25f0772d2748d60f7e4b81bb14d086eba8e8e8efb6dcff8a4ae02"[..])
+            .unwrap();
+        let expected = vec![0u8; 0];
+
+        let output = ecrecover(&input).unwrap();
+
+        assert_eq!(expected, output);
+    }
+
+    #[test]
+    fn invalid_signature() {
+        // Same thing as the geth test, with a corrupt signature
+        let input = FromHex::from_hex(&"38d18acb67d25c8bb9942764b62f18e17054f66a817bd4295423adf9ed98873e000000000000000000000000000000000000000000000000000000000000001bffffffffffffffffffffffffffffffffffffffffffffd4295423adf9ed98873e789d1dd423d25f0772d2748d60f7e4b81bb14d086eba8e8e8efb6dcff8a4ae02"[..])
+            .unwrap();
+
+        let output = ecrecover(&input);
+
+        let error = output.expect_err("Expected an error because of an invalid signature");
+        assert_eq!(InvalidSignature, error);
+
+        // Set s = 0
+        let input = FromHex::from_hex(&"38d18acb67d25c8bb9942764b62f18e17054f66a817bd4295423adf9ed98873e000000000000000000000000000000000000000000000000000000000000001b38d18acb67d25c8bb9942764b62f18e17054f66a817bd4295423adf9ed98873e0000000000000000000000000000000000000000000000000000000000000000"[..])
+            .unwrap();
+
+        let output = ecrecover(&input);
+        let error = output.expect_err("Expected an error because of an invalid signature");
+        assert_eq!(InvalidSignature, error);
+    }
+
+    #[test]
+    fn large_signature() {
+        // Same thing as the geth test, making sure that the s field is
+        // greater than secp256k1n/2.
+        let input = FromHex::from_hex(&"38d18acb67d25c8bb9942764b62f18e17054f66a817bd4295423adf9ed98873e000000000000000000000000000000000000000000000000000000000000001b38d18acb67d25c8bb9942764b62f18e17054f66a817bd4295423adf9ed98873e7fffffffffffffffffffffffffffffff5d576e7357a4501ddfe92f46681b20a0"[..])
+            .unwrap();
+
+        let output = ecrecover(&input).unwrap();
+
+        // The yellow paper says that the signature is invalid if:
+        //
+        // (281) `0 < s < secp256k1n ÷ 2 + 1`
+        //
+        // but this isn't enforced. Check that the contract doesn't
```

Comment:
Can you clarify this into: "isn't enforced for the precompile, only for external transactions." ?